lib/staccato/v4/event.rb

```diff
@@ -1,76 +1,145 @@
-
-      event.class_eval do
-        attr_accessor :tracker, :options
-
-        def_delegators :@options, *event::FIELDS
-
-        # class name, underscored, symbolized
-        def self.event_name
-          @event_name || begin
-            # gsub!(/::/, '/').
-            to_s
-              .dup
-              .split('::')
-              .last
-              .gsub(/([A-Z]+)([A-Z][a-z])/, '\1_\2')
-              .gsub(/([a-z\d])([A-Z])/, '\1_\2')
-              .tr('-', '_')
-              .downcase
-              .to_sym
+      # this module is included into each model hit type
+      #   to share the common behavior required to hit
+      #   the Google Analytics /mp/collect api endpoint
+      def included(klass)
+        klass.class_eval do
+          extend ClassMethods
+          extend Forwardable
+
+          def_delegators :@options, :[], :[]=
+
+          @recommended_params = []
+          fields = defined?(klass::FIELDS) ? klass::FIELDS : []
+          fields.each do |name|
+            param name
           end
-        end
 
-        def name
-          self.class.event_name
+          # In order for user activity to display in standard reports like
+          # Realtime, engagement_time_msec and session_id must be supplied as
+          # part of the params for an event.
+          param :engagement_time_msec
+          param :session_id
+
+          # The time to associate with the event, as a Unix timestamp in
+          # microseconds.  This should only be set to record events that
+          # happened in the past.
+          #
+          # Overrides the tracker timestamp_micros.
+          param :timestamp_micros
         end
+
+        events[klass.event_name] = klass
       end
 
-      events[event.event_name] = event
     end
 
-    def initialize(tracker, options = {})
-      self.tracker = tracker
-      self.options = Staccato::OptionSet.new(options)
+    module ClassMethods
+
+      attr_reader :recommended_params
+
+      def event_name
+        @event_name ||= name
+          .dup
+          .split('::')
+          .last
+          .gsub(/([A-Z]+)([A-Z][a-z])/, '\1_\2')
+          .gsub(/([a-z\d])([A-Z])/, '\1_\2')
+          .tr('-', '_')
+          .downcase
+          .to_sym
+      end
+
+      def param(name)
+        # TODO: prevent assignment of restricted params
+        name = name.to_sym
+        recommended_params << name
+        def_delegator :@options, name
+      end
+
     end
 
-    # return the fields for this hit type
-    # @return [Array] the fields
-    def fields
-      self.class::FIELDS
+    def initialize(**options)
+      @options = Staccato::OptionSet.new(options)
+    end
+
+    def name
+      self.class.event_name
     end
 
     def as_json
-      {
-        name: name,
-        params: event_params
-      }
+      Staccato::V4.encode_params_json({name: name, params: params})
     end
 
-    private
+    def ==(other) other.class == self.class && other.as_json == as_json end
```

Comment:
Is this used? I'm not seeing it, but I could be missing it.

Reply:
Feels like `as_json` is an "expensive" method, but it's only really for the tests, cool.

Reply:
Re: expensive, I should probably add a short-circuit at the front of the boolean expression: `super || ...` to check object identity before checking the contents.

In this case, `as_json` is really just a way to iterate through all of the contents while also compacting away `nil`s.  And that's exactly what's needed for checking equality.